### PR TITLE
nixos/syncthing: add options for declarative device/folder config

### PR DESCRIPTION
**Pull request by Lassulus**
###### Motivation for this change

configure syncthing declaratively in `configuration.nix` (the true way ;) )

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by davidak** [comment]
It would be nice to be able to add a device as Introducer, so it can add the other devices from the network automatically.

**Comment by Lassulus** [comment]
> It would be nice to be able to add a device as Introducer, so it can add the other devices from the network automatically.

Sure, I will add the option in the device submodule.
I have not represented all options from the syncthing API, just the ones I'm using. If you want more: just request it here while this is open.

**Comment by rnhmjoj** [comment]
Nice! I have really wanted something like this, I thought it wasn't possible.
I'll try the PR when I have the chance.

**Comment by mrVanDalo** [comment]
This piece of configuration ends up in the following naming.
```
      folders = {
        podcasts = {
          path = "/home/palo/test-podcasts";
          id = "yvzmx-hcomd";
          devices = [ "workhorse" "schasch" "kruck" ];
        };
      };
```

image

I would be great if the attribute name would become the folder name (like it is for devices).

**Comment by mrVanDalo** [comment]
But apart from the folding naming the whole things works just fine.

There are a few things I would add, for example an `enable` flag for every device and folder entry, 
this way I it's easier for me to define all my folder setups, and enable/disable them on specific computers. But I will create a pull-request once this PR is through (focus on the important parts first).

so it is a :+1:

**Comment by mrVanDalo** [comment]
This seems to create problems.
```
      cert = toString <secrets/syncthing/cert.pem>;
      key = toString <secrets/syncthing/key.pem>;
```
```
error: cannot coerce a set to a string, at /var/src/nixpkgs/nixos/modules/system/boot/systemd-lib.nix:102:10
(use '--show-trace' to show detailed location information)
```

File: nixos/modules/services/networking/syncthing.nix [hunk 1]
```diff
@@ -151,6 +383,20 @@ in {
           RestartForceExitStatus="3 4";
           User = cfg.user;
           Group = cfg.group;
+          ExecStartPre = mkIf (cfg.declarative.cert != null || cfg.declarative.key != null) {
```
Comment (mrVanDalo) [inline nixos/modules/services/networking/syncthing.nix]:
```          ExecStartPre =
            let
              certScript = pkgs.writeShellScriptBin "certScript" (optionalString (cfg.declarative.cert != null)
              ''
                cp ${toString cfg.declarative.cert} ${cfg.configDir}/cert.pem
                chown ${cfg.user}:${cfg.group} ${cfg.configDir}/cert.pem
                chmod 400 ${cfg.configDir}/cert.pem
              '');
              keyScript = pkgs.writeShellScriptBin "keyScript" (optionalString (cfg.declarative.key != null)
              ''
                cp ${toString cfg.declarative.key} ${cfg.configDir}/key.pem
                chown ${cfg.user}:${cfg.group} ${cfg.configDir}/key.pem
                chmod 400 ${cfg.configDir}/key.pem
              '');
            in
            [
              "+${certScript}/bin/certScript"
              "+${keyScript}/bin/keyScript"
            ];
```
Fixing the issue I described, and `ExecStartPre` will be executed with super user rights.

**Review by mrVanDalo — CHANGES_REQUESTED** [review]
The `cert` and `key` argument need to be fixed

File: nixos/modules/services/networking/syncthing.nix [hunk 2]
```diff
@@ -151,6 +383,20 @@ in {
           RestartForceExitStatus="3 4";
           User = cfg.user;
           Group = cfg.group;
+          ExecStartPre = mkIf (cfg.declarative.cert != null || cfg.declarative.key != null)
+            "+${pkgs.writer.writeBash "syncthing-copy-keys" ''
+              ${optionalString (cfg.declarative.cert != null) ''
+                cp ${toString cfg.declarative.cert} ${cfg.configDir}/cert.pem
```
Comment (mrVanDalo) [inline nixos/modules/services/networking/syncthing.nix]:
```
May 06 11:50:03 workout certScript[19729]: chmod: cannot access '/home/palo/.syncthing/.config/syncthing/cert.pem': No such file or directory
May 06 11:50:03 workout certScript[19729]: chown: cannot access '/home/palo/.syncthing/.config/syncthing/cert.pem': No such file or directory
May 06 11:50:03 workout certScript[19729]: cp: cannot create regular file '/home/palo/.syncthing/.config/syncthing/cert.pem': No such file or directory
```

it is not given that `cfg.configDir` exists at this point.

**Review by mrVanDalo — CHANGES_REQUESTED** [review]

File: nixos/modules/services/networking/syncthing.nix [hunk 3]
```diff
@@ -151,6 +383,23 @@ in {
           RestartForceExitStatus="3 4";
           User = cfg.user;
           Group = cfg.group;
+          ExecStartPre = mkIf (cfg.declarative.cert != null || cfg.declarative.key != null)
+            "+${pkgs.writer.writeBash "syncthing-copy-keys" ''
```
Comment (mrVanDalo) [inline nixos/modules/services/networking/syncthing.nix]: `pkgs.writerS.`

**Review by mrVanDalo — CHANGES_REQUESTED** [review]

**Review by mrVanDalo — APPROVED** [review]

**Comment by jabranham** [comment]
A recent syncthing change (I'm assuming this one) made it so that config changes made in the web interface (e.g. adding a device) don't persist across reboots. Is that intended behavior now that declarative config is possible, or should I open an issue?

**Comment by Lassulus** [comment]
Uhm, If you haven't enabled anything with a declarative option that is indeed a bug and I would be happy to fix it. there are 2 options (overrideFolder, overrideDevice) which should only be needed if any device/folder is declarative configured.
For now as a workaround you could try setting:
```
services.syncting.declarative = {
  overrideDevice = false;
  overrideFolder = false;
};
```

**Comment by jabranham** [comment]
> If you haven't enabled anything with a declarative option

I didn't change anything in the syncthing config. This happened after a `nixos-rebuild switch --upgrade` and then reboot.

Thanks for the workaround, I'll have to try it out. Or, perhaps even better, just move to this (awesome) new declarative config.

**Comment by Lassulus** [comment]
Alright, I made a PR for the fix:
https://github.com/NixOS/nixpkgs/pull/62157
sorry for the bug!

**Pull request merged**